### PR TITLE
feat(kafka): support confluent references

### PR DESCRIPTION
**Pull request by anshbansal**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 1]
```diff
@@ -91,6 +91,31 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
             else:
                 self.report.report_dropped(t)
 
+    def get_schema_str_replace_confluent_ref_avro(
+        self, schema: Schema, schema_seen: Dict = None
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/kafka.py:95]: schema_seen: Optional[Dict] = None

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 2]
```diff
@@ -91,6 +91,31 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+    ) -> str:
+        if not schema.references:
+            return schema.schema_str
+
+        if schema_seen is None:
+            schema_seen = dict()
+        schema_str = schema.schema_str
+        for schema_ref in schema.references:
+            ref_subject = schema_ref["subject"]
+            ref_name = schema_ref["name"]
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/kafka.py:105]: move this closer to first usage.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 3]
```diff
@@ -91,6 +91,31 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+            if ref_subject in schema_seen:
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/kafka.py:106]: Looks like `schema_seen` is being used more like a `set` than a `dict`, and is only part of the formal params to facilitate recursive calls. Consider eliminating it from the formal params and use an inner/helper function to facilitate the recursion.

File: metadata-ingestion/src/datahub/ingestion/source/kafka.py [hunk 4]
```diff
@@ -91,6 +91,31 @@ def get_workunits(self) -> Iterable[MetadataWorkUnit]:
...
+                continue
+            tmp_schema = self.schema_registry_client.get_latest_version(ref_subject)
+            schema_seen[ref_subject] = tmp_schema
+            logger.debug(f"ref for {ref_subject} is {tmp_schema.schema.schema_str}")
+            schema_str = schema_str.replace(
+                f'"{ref_name}"',
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/kafka.py:112]: Do the reference names always appear in double quotes? Can they be single quoted? What if this string appears in documentation? Does `confluent_kafka` provider some a parser for us to more deterministically replace the reference schemas names with their schema definitions to generate a well-formed AVRO schema?

Comment (anshbansal) [reply metadata-ingestion/src/datahub/ingestion/source/kafka.py:112]: Avro has to be valid JSON so this should be in double quotes always. `confluent_kafka` does not provide any such method.

**Comment by shirshanka** [comment]
@anshbansal: could you add a unit test for this with a checked in sample schema repo with references?

**Comment by anshbansal** [comment]
@shirshanka Unit test added. Can this be merged now?

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**